### ISSUE TITLE
[BUG] Model keeps on "warming up" state on one of the two macos 26.4 macbooks and gets stuck.

### ISSUE THREAD

**Issue by acalasanzs** [exo-explore/exo#1793]
## Describe the bug

I've noticed that this had only happened on sistematic manner for the Qwen3 235b 4bit mlx model (downloaded directly from exo dashboard). Other models seem to work as for latest main commit. Before that i was on macmon-fix branch (apparently removed from remote) and it worked well for Qwen3 235b 4bit mlx model.

## To Reproduce

1. Connect two macos 26.4 machines through thunderbolt 5 with pipeline and tcp/ip
2. Load Qwen3 235b 4bit mlx model 
3. It keeps saying "Warming up" forever.

## Expected behavior

It should eventually be ready to chat

## Actual behavior

LOGS:
Mac A: runner loaded                                                                                                                
[ 03:23:08.0428PM | INFO    ] Worker plan: StartWarmup                                                                                
[ 03:23:08.0433PM | INFO    ] Starting task StartWarmup(task_id='0a8bf755-c28c-4676-9f36-08b5f54caabd'                                
task_status=<TaskStatus.Pending: 'Pending'> instance_id='81b47d7d-4fd0-4e17-9d5a-3c5fad180afb')                                       
[ 03:23:08.0440PM | INFO    ] runner warming up                                                                                       
[ 03:23:08.0442PM | INFO    ] warming up inference for instance: mlx-community/Qwen3-235B-A22B-Instruct-2507-4bit                     
[ 03:23:08.0583PM | INFO    ] <|im_start|>user                                                                                        
Prompt to warm up the inference engine. Repeat this.<|im_end|>                                                                        
<|im_start|>assistant                                                                                                                 
                                                                                                                                      
[ 03:23:08.0585PM | INFO    ] Using default KV cache                                                                                  
[ 03:23:08.1182PM | INFO    ] Generating warmup tokens                                                                                
[ 03:23:16.1077PM | INFO    ] warmed up by generating 6 tokens                                                                        
[ 03:23:22.4617PM | INFO    ] runner checking for cancellation every 100 tokens                                                       
[ 03:23:22.4620PM | INFO    ] runner initialized in 28.20747709274292 seconds                                                         
[ 03:23:22.4621PM | INFO    ] runner ready Mac B:runner loaded                                                                        
[ 03:23:08.0825PM | INFO    ] Worker plan: StartWarmup                                                                                
[ 03:23:08.0832PM | INFO    ] Starting task StartWarmup(task_id='c92dfbd0-91b5-4178-9b01-81fa0df418a4'                                
task_status=<TaskStatus.Pending: 'Pending'> instance_id='81b47d7d-4fd0-4e17-9d5a-3c5fad180afb')                                       
[ 03:23:08.0841PM | INFO    ] runner warming up                                                                                       
[ 03:23:08.0847PM | INFO    ] warming up inference for instance: mlx-community/Qwen3-235B-A22B-Instruct-2507-4bit                     
[ 03:23:08.0996PM | INFO    ] <|im_start|>user                                                                                        
Prompt to warm up the inference engine. Repeat this.<|im_end|>                                                                        
<|im_start|>assistant                                                                                                                 
                                                                                                                                      
[ 03:23:08.0997PM | INFO    ] Using default KV cache                                                                                  
[ 03:23:08.1002PM | INFO    ] Generating warmup tokens. For some reason one of the machines never gets updated during                 
warm up process and only happens for Qwen3 235b model which stays "warming up" useless

## Environment

- macOS Version: 26.4
- EXO Version: latest main commit as for now which is 2da740c387c7c6b99ebeb61334d1af91ca772c9b hash
- Hardware:
  - Device 1: (e.g., MacBook Pro M4 Pro, 48GB RAM)
  - Device 2: (e.g., Macbook Pro M5 Max, 128GB RAM)
- Interconnection:
  - Thunderbolt 5 cable between Device 1 and 2

## Additional context
I don't have more context than that i've ran claude code with Opus 4.6 and discovered this:
Root Cause                                                                                                                          
                                                                                                                                      
  In pipeline-parallel mode, during warmup is_prefill=True is set, which skips the all_gather in PipelineLastLayer. This means only   
  the last rank gets real logits — all other ranks run lm_head on intermediate hidden states, producing garbage logits.               
                                                                                                                                      
  stream_generate (mlx_lm) checks for EOS tokens after each step (line 710: if token in tokenizer.eos_token_ids: break). The rank with
   garbage logits is very likely to sample an EOS token early and exit the generation loop. Once that rank exits, the other rank hangs
   forever on recv_like waiting for data that will never come.                                                                        
                                                            
  Your logs confirm this: Mac A generated only 6 out of 50 tokens (hit EOS from garbage), then Mac B's recv_like blocks permanently.  
   
  Fix                                                                                                                                 
                                                            
  Ban EOS tokens during warmup so all ranks always generate exactly max_tokens tokens and stay in sync:                               
   
⏺ Update(src/exo/worker/engines/mlx/generator/generate.py)                                                                            
  ⎿  Added 7 lines                                                                        
      334      # Use a default sampler for warmup
      335      sampler = make_sampler(temp=0.0)                                                                                       
      336  
      337 +    # Ban EOS tokens during warmup to prevent early exit.                                                                  
      338 +    # In pipeline-parallel mode, non-last ranks produce garbage logits (intermediate                                
      339 +    # hidden states run through lm_head) and may sample EOS early. If one rank exits                                       
      340 +    # stream_generate while another is still waiting on recv_like, the cluster deadlocks.                               
      341 +    eos_ban = ban_token_ids(list(tokenizer.eos_token_ids))                                                          
      342 +                                                                                                                    
      343      mx_barrier(group)
      344  
      345      logger.info("Generating warmup tokens")
     ...
      357              prefill_step_size=2048,
      358              kv_group_size=KV_GROUP_SIZE,
      359              kv_bits=KV_BITS,
      360 +            logits_processors=[eos_ban],                                                                            
      361          ):
      362              tokens_generated += 1

I didn't verifyied it so maybe it's useless.

**Comment by acalasanzs** [comment]
UPDATE: Claude code's code changes updated on both machines fixes this apparently!

**Comment by Evanev7** [comment]
thanks for the update! looking into this now

**Comment by rltakashige** [comment]
Closing this as resolved, as I am fairly sure #1794 addresses this issue. Please reopen if you run into this again!
